MoltenVK/MoltenVK/GPUObjects/MVKSync.h

```diff
@@ -636,3 +636,56 @@ class MVKMetalCompiler : public MVKBaseObject {
 	std::string _compilerType = "Unknown";
 	MVKPerformanceTracker* _pPerformanceTracker = nullptr;
 };
+
+#pragma mark -
+#pragma mark MVKDeferredOperation
+
+/** Defines the function pointer for each dependent function*/
+union MVKDeferredOperationFunctionPointer
+{
+    // Empty until deferred functions from other extensions have been defined
+    // Planning to use std::functions
+};
+
+/** Manages what function is being deferred*/
+enum MVKDeferredOperationFunctionType
+{
+    // Empty until deferred functions from other extensions have been defined
+};
+
+class MVKDeferredOperation : public MVKVulkanAPIDeviceObject {
+public:
+    /** Returns the Vulkan type of this object. */
+    VkObjectType getVkObjectType() override { return VK_OBJECT_TYPE_DEFERRED_OPERATION_KHR; }
+    
+    /** Returns the debug report object type of this object. */
+    VkDebugReportObjectTypeEXT getVkDebugReportObjectType() override { return VK_DEBUG_REPORT_OBJECT_TYPE_UNKNOWN_EXT; }
+    
+    /** Joins the deferred operation */
+    VkResult join();
+    
+    /** Gets the max number of threads that can execute the deferred operation concurrently*/
+    uint32_t getMaxConcurrency() { return 1; } // Perhaps the number of CPU cores
+    
+    /** Gets the result of the execution of the deferred operation */
+    VkResult getResult() { return operationResult; }
```

Comment:
The functions that set and retrieve the execution state values (eg. `_opearationResult`, `_maxConcurrency`) will almost certainly need a locked mutex, won't they? Some are called from the main thread, some are called by the spawned task thread.

Reply:
Do you think the function pointers need mutexes as well?

Reply:
> Do you think the function pointers need mutexes as well?

From my understanding, the function pointer, type, and parameters, are only modified in `deferOperation()`, after which they become fixed.

If that is the case, from my reading of the spec for this extension, I assume `deferOperation()` will only be called once from the main thread during `vkDoSomethingExpensive()` (from the spec example), well before the function pointer, type, and parameters will be read-only accessed by the threads.

So...if that all makes sense, I don't think we need to mutex those member vars.

But `_opearationResult` and `_maxConcurrency` are clearly intended to be accessible from multiple threads concurrently, so I think we'll definitely need getter and setter functions that mutex those values.